src/main/scala/nl/knaw/dans/lib/error/package.scala

```diff
@@ -76,4 +76,145 @@ package object error {
     }
   }
+
+  implicit class FailFastStream[T](val stream: Stream[Try[T]]) {
+    /**
+     * Evaluates a `Stream` of `Try`s into either:
+     *  - one `Success` with a stream of `T`s or
+     *  - a `Failure` with the first exception contained in the stream
+     *
+     * Note that when the `Stream` encounters a `Failure`, the remaining elements are not evaluated.
+     * The `Failure` is returned immediately.
+     *
+     * @example
+     * {{{
+     *   import nl.knaw.dans.lib.error._
+     *
+     *   import scala.util.Try
+     *
+     *   def f(i: Int) = {
+     *     if (i <= 2) i
+     *     else throw new Exception(s"$$i is larger than 2")
+     *   }
+     *
+     *   val stream: Try[Stream[Int]] = (0 to 5).toStream.map(i => Try(f(i))).failFast
+     *   println(stream)
+     *   // prints: Failure(java.lang.Exception: 3 is larger than 2)
+     * }}}
+     *
+     * @return a evaluated result
+     */
+    def failFast: Try[Stream[T]] = {
+      stream.find(_.isFailure)
+        .map(_.flatMap(s => Failure(new IllegalArgumentException(s"Success should never occur here, but got Success($s)"))))
+        .getOrElse(Success(stream.map(_.get)))
+    }
+  }
+
+  implicit class TryExtensions[T](val t: Try[T]) extends AnyVal {
+    /**
+     * Applies the given side effecting function if and only if this is a `Success`.
+     *
+     * Example:
+     * {{{
+     *   import nl.knaw.dans.lib.error.TryExtensions
+     *
+     *   import scala.util.{Failure, Success, Try}
+     *
+     *   def getFileLength(file: File): Try[Long] =
+     *     if (file.exists) Success(file.length)
+     *     else Failure(new FileNotFoundException())
+     *
+     *   def performSideEffect(size: Long): Unit = println(s"size = $$size")
+     *
+     *   // Fill in existing or non-existing file
+     *   val file = new File("x")
+     *
+     *   getFileLength(file)
+     *     .doIfSuccess(size => performSideEffect(size))
+     * }}}
+     *
+     * @param f the side effecting function to be applied
+     * @return the original `Try`
+     */
+    def doIfSuccess(f: T => Unit): Try[T] = {
+      t match {
+        case success @ Success(x) => Try {
+          f(x)
+          return success
+        }
+        case e => e
+      }
+    }
+
+    /**
+     * Applies the given side effecting `PartialFunction` if and only if this is a `Failure` and
+     * the `Throwable` is defined in the `PartialFunction`.
+     *
+     * Example:
+     * {{{
+     *   import nl.knaw.dans.lib.error.TryExtensions
+     *
+     *   import scala.util.{Failure, Success, Try}
+     *
+     *   def getFileLength(file: File): Try[Long] =
+     *     if (file.exists) Success(file.length)
+     *     else Failure(new FileNotFoundException())
+     *
+     *   // Fill in existing or non-existing file
+     *   val file = new File("x")
+     *
+     *   getFileLength(file)
+     *     .doIfFailure {
+     *       case e: FileNotFoundException => println(e.getMessage)
```

Comment:
You are completely correct! The argument of `doIfFailure` is a `PartialFunction`, meaning that you do not have to specify the function for every type of input. This is the same as in `recover` and `recoverWith`.

> no side effect is executed on another type of exception

indeed, that is also what the code says: if `f isDefinedAt e`, then perform the side effect, otherwise just pass along the original `Try`.